modules/message-lane/src/instant_payments.rs

```diff
@@ -55,34 +65,252 @@ where
+		} else {
+			// If delivery confirmation is submitted by this relayer, let's add confirmation fee
+			// from other relayers to this relayer reward.
+			confirmation_relayer_reward = confirmation_relayer_reward.saturating_add(reward.reward);
+			continue;
+		}
+
+		pay_relayer_reward(currency, relayer_fund_account, &relayer, relayer_reward);
+	}
+
+	// finally - pay reward to confirmation relayer
+	pay_relayer_reward(
+		currency,
+		relayer_fund_account,
+		confirmation_relayer,
+		confirmation_relayer_reward,
+	);
+}
+
+/// Transfer funds from relayers fund account to given relayer.
+fn pay_relayer_reward<AccountId, Balance, Currency>(
+	currency: &mut Currency,
+	relayer_fund_account: &AccountId,
+	relayer_account: &AccountId,
+	reward: Balance,
+) where
+	AccountId: Debug,
+	Balance: Debug + Zero + Copy,
+	Currency: InstantCurrency<AccountId, Balance>,
+{
+	if reward.is_zero() {
+		return;
+	}
+
+	let pay_result = currency.transfer(relayer_fund_account, relayer_account, reward);
 
-		// we can't actually do anything here, because rewards are paid as a part of unrelated transaction
-		match pay_result {
-			Ok(_) => frame_support::debug::trace!(
-				target: "runtime",
-				"Rewarded relayer {:?} with {:?}",
-				relayer,
-				reward,
-			),
-			Err(error) => frame_support::debug::trace!(
-				target: "runtime",
-				"Failed to pay relayer {:?} reward {:?}: {:?}",
-				relayer,
-				reward,
-				error,
-			),
+	// we can't actually do anything here, because rewards are paid as a part of unrelated transaction
+	match pay_result {
+		Ok(_) => frame_support::debug::trace!(
+			target: "runtime",
+			"Rewarded relayer {:?} with {:?}",
+			relayer_account,
+			reward,
+		),
+		Err(error) => frame_support::debug::trace!(
+			target: "runtime",
+			"Failed to pay relayer {:?} reward {:?}: {:?}",
+			relayer_account,
+			reward,
+			error,
+		),
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use bp_message_lane::source_chain::RelayerRewards;
+	use std::{cell::RefCell, collections::HashMap};
+
+	type TestAccountId = u64;
+	type TestBalance = u64;
+
+	const RELAYERS_FUND_ACCOUNT: TestAccountId = 10;
+
+	struct TestCurrency {
+		balances: RefCell<HashMap<TestAccountId, TestBalance>>,
+	}
+
+	impl TestCurrency {
+		fn new() -> Self {
+			TestCurrency {
+				balances: RefCell::new(vec![(RELAYERS_FUND_ACCOUNT, 1_000)].into_iter().collect()),
+			}
+		}
+	}
+
+	impl InstantCurrency<TestAccountId, TestBalance> for TestCurrency {
+		fn transfer(&mut self, source: &TestAccountId, dest: &TestAccountId, value: TestBalance) -> DispatchResult {
+			let mut balances = self.balances.borrow_mut();
+			*balances.entry(*source).or_default() -= value;
+			*balances.entry(*dest).or_default() += value;
+			Ok(())
 		}
 	}
+
+	#[test]
+	fn confirmation_relayer_is_rewarded_if_it_has_also_delivered_messages() {
+		let mut currency = TestCurrency::new();
+		pay_relayers_rewards(
```

Comment:
I found myself flipping back and forth between the tests and code bit. Some intermediate variables with the relayer_ids and fees for example would help in making the test a bit more legible in a "standalone" way